posts/2022-06-01-RLS-deprecation.md

```diff
@@ -0,0 +1,30 @@
+---
+layout: post
+title: "RLS Deprecation"
+author: The Rust Dev Tools Team
+---
+
+The Rust Language Server (RLS) is being deprecated in favor of [rust-analyzer](https://rust-analyzer.github.io/).
+Current users of RLS should migrate to using rust-analyzer instead.
+Builds of RLS will continue to be released until at least the Rust 1.64 release (2022-09-22), after which no new releases will be made.
+This timeline may change if any issues arise.
+
+RLS is an implementation of the [Language Server Protocol](https://microsoft.github.io/language-server-protocol/) (LSP) which provides enhanced features with any editor that supports the protocol, such as code-checking and refactoring.
+RLS was introduced by [RFC 1317](https://rust-lang.github.io/rfcs/1317-ide.html) and development was very active from 2016 through 2019.
+However, the architecture of RLS has several limitations that can make it difficult to provide low-latency and high-quality responses needed for an interactive environment.
+
+Development of rust-analyzer began near the beginning of 2018 to provide an alternate LSP implementation for Rust.
+rust-analyzer uses a fundamentally different approach that does not rely on using `rustc`.
+In [RFC 2912](https://rust-lang.github.io/rfcs/2912-rust-analyzer.html) rust-analyzer was adopted as the official replacement for RLS.
+
+How you migrate to rust-analyzer will depend on which editor you are using.
+If you are using VSCode, you should uninstall the `rust-lang.rust` extension and install the official [`rust-lang.rust-analyzer` extension](https://marketplace.visualstudio.com/items?itemName=rust-lang.rust-analyzer).
+For other editors, please consult the [rust-analyzer manual](https://rust-analyzer.github.io/manual.html) for instructions on how to install it.
```

Comment:
I don't know. I opened https://github.com/rust-lang/rust-analyzer/issues/12432 to track that question.

@cuviper Do you know if there are users that would benefit from a stable-channel release? I'm not familiar with how current non-VSCode users handle it.

Reply:
I'll respond further on that issue, but for myself as a Vim user, I would appreciate a stable release. I used to manually download the weekly rust-analyzer releases, but lately I've taken to using the nightly component, symlinked into my global `PATH` so it's usable for all toolchain versions. I have nightly installed anyway, but I don't think that's a great solution for someone who only wants the stable toolchain.

As a distro maintainer, I would also like to build and package rust-analyzer in my rpm builds, but that's risky if stable builds aren't really a thing here in the upstream project. I know some people insist that distro compilers are only for building the distro itself, so you wouldn't need LSP, but that view is not true of GCC, and I don't think that should be true of Rust either.

Reply:
@cuviper Is it OK to resolve this conversation as being tracked in the linked issue?  Or do you feel that having it available on stable is a blocker to deprecation?

Reply:
Maybe I missed something, how is shipping on more platforms related to stable? We already have rust-analyzer on rustup nightly, does that somehow support fewer platforms than stable?

Reply:
@flodiebold nightly does have all the same platforms as stable -- I meant that in comparison to the repo [releases](https://github.com/rust-lang/rust-analyzer/releases), and I don't think we should consider the nightly channel to be any sort of official release. Since we have had that build in nightly for some time now, AFAIK without trouble, this is why I think it shouldn't be too much of a technical lift to let that go out to stable.

Reply:
That seems like the main argument is more that rust-analyzer should have "stable" releases rather than about the platforms it's available on then. The current repo releases, and the published VSCode extension, aren't any more bug-free or stable than what's published in rustup (except by being potentially more recent, which having a stable channel would make even worse).

Reply:
But you do have to install the nightly toolchain to get that rustup component, and that will itself be built with the same (non-release) nightly `rustc`. Also, if being recent was _inherently_ good, then we would just tell everyone to use nightly. The release train does serve a purpose, if only to let things settle a bit.

Reply:
> But you do have to install the nightly toolchain to get that rustup component

There is the idea of introducing an `--profile none` to rustup which would allow `rustup toolchain install nightly --profile none --component rust-analyzer` to only install rust-analyzer and not rustc or anything else.

> Also, if being recent was inherently good, then we would just tell everyone to use nightly for everything.

The rust-analyzer available on the nightly rustc channel is the stable rust-analyzer, not the nightly rust-analyzer.

> The release train does serve a purpose, if only to let things settle a bit.

Rust-analyzer doesn't have a release train at all. Getting any benefit from rust-analyzer on the stable rustc channel would require introducing a release train to rust-analyzer.

Reply:
> Rust-analyzer doesn't have a release train at all.

Sure, but I'm saying that it *should* follow the release train for distribution with Rust, rather than being exception to everything else that the project ships. Whether they also continue with faster non-train releases is up to them.

Reply:
@cuviper With https://github.com/rust-lang/rust/pull/98640 to include it on stable, does that resolve your concern here?

Reply:
Yes, thanks!